rust/osm2lanes/src/transform/tags_to_lanes/lane_count.rs

```diff
@@ -0,0 +1,366 @@
+use crate::locale::{DrivingSide, Locale};
+use crate::road::Designated;
+use crate::tag::{TagKey, Tags};
+use crate::transform::tags_to_lanes::{Infer, Oneway, RoadBuilder};
+use crate::transform::{RoadWarnings, TagsToLanesMsg};
+
+const LANES: TagKey = TagKey::from("lanes");
+const ONEWAY: TagKey = TagKey::from("oneway");
+
+#[derive(Debug)]
+pub(in crate::transform::tags_to_lanes) struct Counts {
+    pub lanes: Infer<usize>,
+    pub forward: Infer<usize>,
+    pub backward: Infer<usize>,
+    pub both_ways: Infer<usize>,
+}
+
+impl Counts {
+    /// Parses and validates the `lanes` scheme (which excludes parking lanes, bike lanes, etc.).
+    /// See <https://wiki.openstreetmap.org/wiki/Key:lanes>.
+    ///
+    /// Uses data from tags like `lanes[:{forward,both_ways,backward}]=*` and `centre_turn_lane=yes`.
+    /// Tags with pipe separated access strings are also used.
+    /// Lanes already existing in `road` used to determine the minima (e.g. for busway).
+    #[allow(
+        clippy::integer_arithmetic,
+        clippy::integer_division,
+        clippy::too_many_lines
+    )]
+    pub(in crate::transform::tags_to_lanes) fn new(
+        tags: &Tags,
+        road: &RoadBuilder,
+        locale: &Locale,
+        warnings: &mut RoadWarnings,
+    ) -> Self {
+        // TODO, crosscheck access against tagged
+        let access = Self::get_counts_by_access(tags, warnings);
+
+        // Pre-existing lanes
+        let pre_forward = road
+            .forward_lanes
+            .iter()
+            .filter(|l| {
+                matches!(
+                    l.designated.some(),
+                    Some(Designated::Bus | Designated::Motor)
+                )
+            })
+            .count();
+        let pre_backward = road
+            .backward_lanes
+            .iter()
+            .filter(|l| {
+                matches!(
+                    l.designated.some(),
+                    Some(Designated::Bus | Designated::Motor)
+                )
+            })
+            .count();
+
+        // The tags for this schema (which we will validate).
+        let tagged_lanes: Option<usize> = tags
+            .get_parsed(&LANES, warnings)
+            .or_else(|| access.lanes.some());
```

Comment:
This is no longer always the tagged value for the lanes key ("`tagged_lanes`"), but sometimes an implied count from access tags, but we can't tell.

Reply:
indeed, so I live all the "hacks" in this monolithic function. this is the part I would totally refactor over time.

Reply:
If that is the case, I think don't name it `Counts::new`, name it `calculate_and_guess_lane_counts(tags, road, ...) -> Counts` or something fittingly "hacky" :)

I found that converting to `Infer` as soon as possible, then using `.some()` if you don't care what you're getting, otherwise you can be smart about reconciling a default with a calculated value etc. I decided against chaining the logic like this, because it turns out if you are combining two sources of the same value, you want to do validation anyway, and having `Infer`s for the is useful.